### PR TITLE
Serialization helpers for STL types WIP

### PR DESCRIPTION
**Pull request by breznak**
Aims to use StlIo.hpp to help serialization

### CODE REVIEW

File: src/nupic/types/Serializable.hpp [hunk 1]
```diff
@@ -33,9 +33,11 @@
 
 #include <iostream>
 #include <fstream>
+
 #include <nupic/os/Directory.hpp>
 #include <nupic/os/Path.hpp>
 #include <nupic/os/ImportFilesystem.hpp>
+#include <nupic/utils/StlIo.hpp> // use these to help you serialize std::containers
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: I think all Serializable objects (those that have Serializable as base class) must override both save( ) and load( ).  A C++ class does not have a way to look at itself and see what variables are in the class ... at least not yet.  This is called Reflection.  See https://stackoverflow.com/questions/41453/how-can-i-add-reflection-to-a-c-application


Also, I think you need to make the base class public so other classes can call Serializable member functions.
```
class A: public Serializable {  }
```

File: src/nupic/types/Serializable.hpp [hunk 2]
```diff
@@ -72,11 +74,21 @@ class Serializable {
   virtual void save(std::ostream &stream) const = 0;
   virtual void load(std::istream &stream) = 0;
 
+  // Serializable overrides stream operators <<, >> 
+  std::ostream &operator<<(std::ostream &outStream) {
+    this->save(outStream);
+    return outStream;
+  }
+
+  std::istream &operator>>(std::istream &inStream) {
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/types/Serializable.hpp]: I tried to do that but ran into some sort of problem and backed it out.

Remember we have two types of serialization.   One type using Serializable class to be read by the load( ) which is mostly binary for speed.  The other type is to be readable by humans which must be text for debugging.  We cannot have the << operator doing both.   I sort of prefer that the << operator be used for the human readable serialization but we should make it consistent.

File: src/nupic/utils/StlIo.hpp [hunk 3]
```diff
@@ -32,126 +32,67 @@
...
+  cout << "YYY1 " << sz << endl;
+
+  if(sz > 0) {    
+    objTo.resize(sz);
+    const auto begin = std::begin(objTo);
+    typedef typename Container::value_type value_type;
+    cout << "YYY2 " << sizeof(value_type) << endl;
+    char* raw = (char*)&*begin;
+    cout << "YYY 3 " << in_stream.tellg() << endl;
+    in_stream.read(raw, (std::streamsize)sz * sizeof(value_type));
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/utils/StlIo.hpp]: Do we need this function?  I think it is trying to load an arbitrary object and that would probably not work if the object contains any pointers.   If this is trying to work with a vector, it can get the destination pointer by getting the address of the first element.  That works because the vector's buffer is contiguous.  But that does not work with std::set or std::map etc. because each element's pair is allocated independently.

File: src/test/unit/utils/SerializationSTLTest.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+class Dummy : public Serializable {
+  public:	
+  int i = 3;
+  vector<int> v = vector<int>{1, 2, 3};
+
+  //Serializable interface
+  void save(ostream &o) const override {
+    o << "DUMMY ";
+    o << i << " ";
+//!    o << v << " ";
```
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/test/unit/utils/SerializationSTLTest.cpp]: Yes, this is a good objective.  
Actually, the vector binary serialization did work before but the logic was hard to follow.

The best way to do this is to provide binary serializations for each of the STL classes that use save( ) and load( ) and then make another set of human readable serializations for each of the STL classes which will use the  << and >> operators.

Remember that the STL classes could be nested.  How do we serialize something like `vector<std::deque<std::map<std::string>>>`  ?

Now, this is a common problem.  I wonder if someone already did this someplace on the web?

File: src/nupic/utils/StlIo.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/utils/StlIo.hpp]: > But that does not work with std::set or std::map etc. because each element's pair is allocated independently.

perfect, so I'll remove the binary_save, binary_load (as only works for vectors!), and I'll make work the StlIo::vector_save() which is "human readable, text" format.

File: src/nupic/utils/StlIo.hpp [hunk 5]
```diff
@@ -184,17 +120,16 @@ inline std::istream &operator>>(std::istream &in_stream, std::pair<T1, T2> &p) {
...
 //--------------------------------------------------------------------------------
 /**
  * Partial specialization of above functor for primitive types.
  */
 template <typename T> struct vector_loader<T, true> {
-  inline void load(size_t n, std::istream &in_stream, std::vector<T> &v) {
-      for (size_t i = 0; i != n; ++i)
-        in_stream >> v[i];
+  inline void load(const size_t n, std::istream &in_stream, std::vector<T> &v) {
+      for (size_t i = 0; i != n; ++i)  in_stream >> v[i];
```
[the remark this answers is not in the document]
Comment (breznak) [reply src/nupic/utils/StlIo.hpp]: binary_save to be removed, this will be used!

File: src/test/unit/utils/SerializationSTLTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (breznak) [reply src/test/unit/utils/SerializationSTLTest.cpp]: > emember that the STL classes could be nested. How do we serialize something like vector<std::deque<std::map<std::string>>> ?

this is why I wanted to go with the binary serialization, I had hoped all STL Containers use continuous memory, so the binary version would just dump/load whatever is inside. 

>  That works because the vector's buffer is contiguous. But that does not work with std::set or std::map etc. because each element's pair is allocated independently. ~DK

But you say it's not true (only for vectors). 

> Yes, this is a good objective.

this is why I wanted to override `operator<<`, as it's used only at a few places for "print", and print() could be used there. While when writing serialization, we stream all primitive types with `<<`, so I wanted to provide that for STL too.

File: src/nupic/utils/StlIo.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkeeney) [reply src/nupic/utils/StlIo.hpp]: But for use in serialization with save( ) and load( ) we should have a binary stream for a vector. Do we have one someplace?

**Comment by dkeeney** [comment]
I found this serialization library   http://uscilab.github.io/cereal/index.html
Its C++11, header only.
Don't want to get into another "captn Proto" but its probably worth looking at.

**Comment by breznak** [comment]
> Don't want to get into another "captn Proto" but its probably worth looking at. 

yes, we should be cautious with that. But cereal does look pretty lightweight and quite fit for the use. Thanks! 
I think an important decision factor should be how easy it is to serialize custom/unsupported datastructures.

**Comment by breznak** [comment]
Maybe changes in this PR shouldn't be so drastical, as I quite like the look of Cereal, so we might be obsoleting StlIo functionality someday.

**Comment by dkeeney** [comment]
As I look at StlIo, I am beginning to think something simple would work.  Cereal might be overkill.  We don't need to serialize arbitrary structures....just STL classes.  But we do need to do two different types of serialization...one binary for speed and one text for readability.

I am thinking along the lines of something like:
```
/** @file STL IO
 * This file contains functions to print/restore as readable text
 * and save/load as binary most of the major STL data structures.
 * There are two types of streaming:
 *      TEXT   - This uses operator<< and operator>> to write and read.
 *      BINARY - This uses the save() and load() functions to write and read.
 * This code attempts to provide both capability.
 *
 * This include allows streaming of the following STL structures:
 *    std::pair<K,T>
 *    std::vector<T>
 *    std::map<K,T>
 *    std::multimap<K,T>
 *    std::set<T>
 *    std::multiset<K,T>
 *    std::list<T>
 *    std::dque<T>
 *    std::stack<T>
 *    std::queue<T>
 *    std::array<N,T>
 *    std::shared_ptr<T>
 *    
 * Rules:
 *   1) The <T> can be any container in the list, any fundamental data
 *      type, or any subclass of Serializable.
 *   2) Cannot handle types that are raw pointers unless wrapped
 *      in a std::array<N,T> so that the number of elements are known.
 *   3) std::shared_ptr<T> must not point to an array. The number of
 *      elements is unknown.
 *   4) Data types can be 'nested' as long as all data types are of the
 *      types in 1, 2, or 3.
 *
 * Usage:
 *   For text streaming, use the operator<< and operator>> overloading
 *   but for binary serialization, use save( ) and load( ).
 *   For example:
 *       std::ostream f;
 *       std::map<std::string name, shared_ptr<Region>> region_map;
 * 
 *       f << region_map;      // get a readable serialization.
 *
 *       save(f, region_map);  // for the binary serialization.
 *   then restoring it with:
 *       load(f, region_map);
 *       
 *   This should handle nested cases like 
 *       map<vector<list<Link&>>> linkmap;
*       f << linkmap;     // for a text serialization
*   And going the other way,
*       f >> linkmap;
 */
```

I can take over this PR if you are not interested in this.

**Comment by breznak** [comment]
I'll be happy if you want to continue here. 

> serialization...one binary for speed and one text for readability. 

Do we? How fast/slow serialization actually is? It'd be nice if the plain textual, human-readable serialization was enough. 

> This include allows streaming of the following STL structures:

for starters, smart_pointer, vector, set, map, array should be enough.

**Comment by dkeeney** [comment]
> Do we? How fast/slow serialization actually is? It'd be nice if the plain textual, human-readable serialization was enough.

I just have a feeling it is not going to be fast enough. But I will not know until we try it with 1000 SP,TP pairs.  Ok, I will do the text first and if we need to we can add the binary later.

> for starters, smart_pointer, vector, set, map, array should be enough.

Ok, it should not be hard to extend it to add others as we find a need.  Oh, will need std::pair in able to do the map.

**Comment by dkeeney** [comment]
This PR is far from finished.  I will be away from my computer for 5 days so there will not be much progress for a while.

What I have so far are routines to handle both >> and << streaming for many of the STL structures.

But nothing is for free.... 

1) It turns out that this is highly dependent on the namespace.  The streaming routines in StlIo.hpp and the objects being streamed must be in the same namespace.  StIo.hpp is all in the nupic:: namespace and so is all of the NetworkAPI.  But the algorithms are not.  namespace nupic::algorithms::temporal_memory  is not the same as nupic namespace and therefore streaming a vector does not work.  Here is a case where a more specific namespace is a bad thing.
2) There is a work-around for cross-namespace operation.  Rather than streaming with `f << vector_obj; ` use the syntax  
```
          nupic::operator<<(outStream, _prevInfPatterns);
          nupic::operator>>(inStream,  _prevInfPatterns);
```
3) Nested STL structures do not work.  I suspect that it is somehow due to the namespace thing but not sure yet.
4) I have not finished working on the std::map.  It needs a few more tweaks to get the pairs to work.
5) I did solve the problem of streaming vectors containing std::string objects that contain spaces.
6) Our save( ) and load( ) functions don't play well with this.  It would be nice to also be able to use these with STL objects as well.   Perhaps:  
```
save(std::ostream& f, const std::vector<T> v);
load(std::istream& f, std::vector<T> v);
```

**Comment by dkeeney** [comment]
Here is a link to a post that explains why cross-namespace streams will not work:
https://stackoverflow.com/questions/5195512/namespaces-and-operator-resolution

Oh, and putting the routines in StlIo.hpp in the default namespace does not work either.

Another thing I found was that the SDR class does not support >> streaming.  So if I serialize a vector containing SDR objects using << it cannot be deserialized with >> streaming.

As I stated before, I am not comfortable with using >> and << for both textual display and for archive save and load.  I still think the archive save and load needs to be binary or at least more compact.

**Comment by ctrl-z-9000-times** [comment]
> Another thing I found was that the SDR class does not support >> streaming. So if I serialize a vector containing SDR objects using << it cannot be deserialized with >> streaming.

Adding >> for sdr is a great idea. I can put this on my list of thing to do.

**Comment by ctrl-z-9000-times** [comment]
> I found this serialization library http://uscilab.github.io/cereal/index.html
> Its C++11, header only.
> Don't want to get into another "captn Proto" but its probably worth looking at.

I looked into the **cereal** library and I fully recommend moving forward with it as a replacement for our current tool set.
* Features: It does everything we want and more.
* Style: It's well documented, tested, the API is well thought out.

### SDR TechDemo Notes:
As a proof of concept I made the SDR class use the cereal library.  The purpose of this exercise was *not* to make a useful implementation, but rather to evaluate & demonstrate this technology.

I cloned their git repo and checked out the latest tag (v1.2.2).
I used a symlink to get it into the CMake includes.
- `nupic.cpp/external/common/include/cereal -> cereal/include/cereal/`

Include cereal in Sdr.hpp
```C++
// Cereal also uses the names save/load"which conflicts with our existing
// serialization methods.  Cereal allows renaming the methods.
#define CEREAL_SAVE_FUNCTION_NAME save2
#define CEREAL_LOAD_FUNCTION_NAME load2

#include <cereal/macros.hpp>
#include <cereal/archives/binary.hpp>
#include <cereal/types/memory.hpp>
#include <cereal/cereal.hpp>
#include <cereal/access.hpp>
#include <cereal/types/vector.hpp>
```

Add the new save2 / load2 methods to the SDR class definition:
```C++
friend class cereal::access; // Let cereal use our private variables.

template<class Archive>
void save2(Archive & archive) const
{
    archive( dimensions, getSparse() );
}

template<class Archive>
void load2(Archive & archive)
{
    archive( dimensions_, sparse_ );
    initialize( dimensions_ );
    setSparseInplace();
}
```

For testing I replaced the implementation of the current save/load API with calls to **cereal**
The unit tests for SDR serialization pass!
```C++
#include <cereal/archives/binary.hpp>
void SparseDistributedRepresentation::save(std::ostream &outStream) const {
        cereal::BinaryOutputArchive archive( outStream );
        archive( *this );
}

void SparseDistributedRepresentation::load(std::istream &inStream) {
        cereal::BinaryInputArchive archive( inStream );
        archive( *this );
}
```

**Comment by breznak** [comment]
> The unit tests for SDR serialization pass! 

:+1: :+1: very nice example! 

> Add the new save2 / load2 methods to the SDR class definition

It should really boild down to how much hassle is to implement the new (serialization library) API into our code.
If it's reasonably simple, I'm all in, as the serialization is not a key focus of our project, it'd be ok to leave it to specialized project.

**Comment by dkeeney** [comment]
This looks good to me to. 
- can we tie it to the .py serialization?  For example if we serialize Network can we get it to also serialize the registered .py regions along with the C++ modules?

- does it have support for nested STL containers like vector, map, etc.

**Comment by ctrl-z-9000-times** [comment]
> does it have support for nested STL containers like vector, map, etc.

Yes, the project page has an example of serializing this example
`std::map<std::string, std::vector<std::complex<float>>> data;`

**Comment by ctrl-z-9000-times** [comment]
> can we tie it to the .py serialization? For example if we serialize Network can we get it to also serialize the registered .py regions along with the C++ modules?

No, this will not serialize arbitrary python data such as `py::object`. For python we should instead use the built in `pickle` module to serialize, and then save the result inside of  cereal's archives.  I think this scheme could be implemented once and work for all python modules.

**Comment by ctrl-z-9000-times** [comment]
> It should really boil down to how much hassle is to implement the new (serialization library) API into our code.
> If it's reasonably simple, I'm all in, as the serialization is not a key focus of our project, it'd be ok to leave it to specialized project.

My thoughts exactly.  Serialization is necessary and annoying.  I think that with a small initial investment in setup and in our training (learning to use this new library), that we could have a significantly better result than what we're planning in this PR.

**Comment by ctrl-z-9000-times** [comment]
> does it have support for nested STL containers like vector, map, etc.

I was curious and I tried a more complex example:  `vector<SDR>`
This works!

`vector<SDR*>` will not work.
`vector<smart_ptr<SDR>>` should work.

**Comment by dkeeney** [comment]
Considering the conflict problems with namespaces I think we should put this on hold or maybe not do it and instead implement the Cereal package.

**Comment by breznak** [comment]
> Considering the conflict problems with namespaces I think we should put this on hold or maybe not do it and instead implement the Cereal package. 

Closing this as obsoleted by Cereal serialization in #358 and others. What should eventually be done is removal of StlIo.hpp
